File: xxhoeckyxx_Test/Zahlsys.java
import java.util.Scanner;

public class Zahlsys {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        // Anzeige der Programmnutzung, wenn nicht genügend Argumente vorhanden sind
        if (args.length < 1) {
            showUsage(args[0]);
            System.exit(1);
        }

        // Initialisierung von Flags und Dezimalzahl
        boolean[] flags = new boolean[3];
        int[] decimal = new int[1];

        // Analyse der Befehlszeilenoptionen
        if (!parseOptions(args, flags, decimal, scanner)) {
            System.exit(1); // Beende das Programm mit Fehlercode 1 bei einem Analysefehler
        }

        // Falls keine Basisoptionen angegeben sind, verwenden Sie alle drei
        if (!flags[0] && !flags[1] && !flags[2]) {
            flags[0] = flags[1] = flags[2] = true;
        }

        // System.out.println("Dual: " + flags[0] + "\nOctal: " + flags[1] + "\nHex: " + flags[2]);

        // Anzeige der Ergebnisse
        printResults(flags[0], flags[1], flags[2], decimal[0]);
    }

    // Funktion zum Analysieren von Zeichenoptionen
    static boolean parseCharacter(String option, boolean[] flags) {
        for (int j = 2; j < option.length(); j++) {
            switch (option.charAt(j)) {
                case 'b':
                    flags[0] = true;
                    break;
                case 'o':
                    flags[1] = true;
                    break;
                case 'h':
                    flags[2] = true;
                    break;
                default:
                    System.err.println("....... Unerlaubte Option " + option + "\n\n");
                    return false;
            }
        }
        return true;
    }

    // Funktion zum Analysieren von Dezimalzahlen mit dem Scanner
    static boolean parseNumber(String option, int[] decimal, Scanner scanner) {
        try {
            decimal[0] = Integer.parseInt(option);
            return true;
        } catch (NumberFormatException e) {
            System.err.println("Fehler: Ungültiger Dezimalwert '" + option + "'\n");
            return false;
        }
    }

    // Hauptfunktion zum Analysieren der Befehlszeilenoptionen
    static boolean parseOptions(String[] args, boolean[] flags, int[] decimal, Scanner scanner) {
        int numberCount = 0;
        int detectFail = 0;

        for (int i = 0; i < args.length; i++) {
            if (args[i].startsWith("--")) {
                // Verarbeitung von Optionen, die mit '--' beginnen
                if (!parseCharacter(args[i], flags)) {
                    showUsage(args[0]);
                    detectFail++;
                }
            } else if (numberCount != 0) {
                System.err.println("....... Es muss genau eine Zahl angegeben sein\n");
                detectFail++;
            } else {
                // Verarbeitung der Dezimalzahl, die nicht mit '-' beginnt
                if (!parseNumber(args[i], decimal, scanner)) {
                    showUsage(args[0]);
                    detectFail++;
                }
                numberCount++;
            }
        }

        return detectFail == 0;
    }

    // Funktion zur Anzeige der Ergebnisse je nach gewählter Basis
    static void printResults(boolean binary, boolean octal, boolean hex, int decimal) {
        System.out.println("           Dezimal: " + decimal + "\n");

        if (binary) {
            // Konvertierung und Anzeige der Binärzahl
            int[] binList = new int[32];
            for (int i = 31; i >= 0; i--) {
                binList[31 - i] = (decimal >> i) & 1;
            }
            System.out.print("              Dual: ");
            for (int i = 0; i < 32; i++) {
                if (i % 8 == 0 && i > 0) {
                    System.out.print(" ");
                }
                System.out.print(binList[i]);
            }
            System.out.println();
        }
        if (octal) {
            // Anzeige der Oktalzahl
            System.out.println("\n             Octal: " + Integer.toOctalString(decimal));
        }
        if (hex) {
            // Anzeige der Hexadezimalzahl
            System.out.println("       Hexadezimal: " + Integer.toHexString(decimal));
        }
    }

    // Funktion zur Anzeige der Programmnutzung
    static boolean showUsage(String programName) {
        System.out.println("usage: " + programName + " [--boh] zahl");
        return false;
    }
}

File: xxhoeckyxx_Test/zahlsys.cpp
#include <iostream>
#include <cstdlib>
#include <cctype>

// Funktion zur Anzeige der Programmnutzung
bool showUsage(const char *programName)
{
    std::cout << "usage: " << programName << " [--boh] zahl\n";
    return false;
}

// Funktion zum Analysieren von Zeichenoptionen
bool parseCharacter(const char *option, bool &binary, bool &octal, bool &hex)
{
    for (int j = 2; option[j] != '\0'; j++)
    {
        switch (option[j])
        {
        case 'b':
            binary = true;
            break;
        case 'o':
            octal = true;
            break;
        case 'h':
            hex = true;
            break;
        default:
            std::cerr << "....... Unerlaubte Option " << option << "\n\n";
            return false;
        }
    }

    return true;
}

// Funktion zum Analysieren von Dezimalzahlen
bool parseNumber(const char *option, int &decimal)
{
    int decimalSet = 0;
    while (decimalSet == 0)
    {
        if (option[0] == '-' && isdigit(option[1]))
        {
            if (sscanf(option, "%d", &decimal) != 1)
            {
                std::cerr << "Fehler: Ungültiger Dezimalwert '" << option << "'\n";
                return false;
            }
            else
            {
                decimalSet = 1;
            }
        }
        else if (isdigit(option[0]))
        {
            if (sscanf(option, "%d", &decimal) != 1)
            {
                std::cerr << "Fehler: Ungültiger Dezimalwert '" << option << "'\n";
                return false;
            }
            else
            {
                decimalSet = 1;
            }
        }
        else
        {
            std::cerr << "Fehler: Unerwartetes '-' im Dezimalwert '" << option << "'\n";
            return false;
        }
    }

    return true;
}

// Hauptfunktion zum Analysieren der Befehlszeilenoptionen
bool parseOptions(int argc, char *argv[], bool &binary, bool &octal, bool &hex, int &decimal)
{
    int numberCount = 0;
    int detectFail = 0;

    for (int i = 1; i < argc; i++)
    {
        if (argv[i][0] == '-' && argv[i][1] == '-')
        {
            // Verarbeitung von Optionen, die mit '--' beginnen
            if (!parseCharacter(argv[i], binary, octal, hex))
            {
                showUsage(argv[0]);
                detectFail++;
            }
        }
        else if (numberCount != 0)
        {
            std::cerr << "....... Es muss genau eine Zahl angegeben sein\n";
            detectFail++;
        }
        else
        {
            // Verarbeitung der Dezimalzahl, die nicht mit '-' beginnt
            if (!parseNumber(argv[i], decimal))
            {
                showUsage(argv[0]);
                detectFail++;
            }
            numberCount++;
        }
    }

    if (detectFail > 0)
    {
        return false;
    }
    else
    {
        return true;
    }
}

// Funktion zur Anzeige der Ergebnisse je nach gewählter Basis
void printResults(bool binary, bool octal, bool hex, int decimal)
{
    std::cout << "           Dezimal: " << decimal << "\n\n";

    if (binary)
    {
        // Konvertierung und Anzeige der Binärzahl
        int bin_list[32] = {0};
        for (int i = 31; i >= 0; i--)
        {
            bin_list[31 - i] = (decimal >> i) & 1;
        }
        std::cout << "              Dual: ";
        for (int i = 0; i < 32; i++)
        {
            if (i % 8 == 0 && i > 0)
            {
                std::cout << " ";
            }
            std::cout << bin_list[i];
        }
        std::cout << "\n";
    }
    if (octal)
    {
        // Anzeige der Oktalzahl
        std::cout << "\n             Octal: " << std::oct << decimal << "\n";
    }
    if (hex)
    {
        // Anzeige der Hexadezimalzahl
        std::cout << "       Hexadezimal: " << std::hex << decimal << "\n";
    }
}

int main(int argc, char *argv[])
{
    if (argc < 2)
    {
        // Anzeige der Programmnutzung, wenn nicht genügend Argumente vorhanden sind
        return showUsage(argv[0]);
    }

    // Initialisierung von Flags und Dezimalzahl
    bool binary = false;
    bool octal = false;
    bool hex = false;
    int decimal = 0;

    // Debugging für argc und argv
    // std::cout << "Argument Count: " << argc << "\n";
    // for (int i = 0; i < argc; i++)
    // {
    //     std::cout << "Argument " << i << ": " << argv[i] << "\n";
    //     std::cout << "SizeOf Argument " << i << ": " << sizeof(argv[i]) << "\n";
    //     for (size_t j = 0; j < sizeof(argv[i]); j++)
    //     {
    //         std::cout << "Character " << j << ": " << argv[i][j] << "\n";
    //         std::cout << "SizeOf Character " << j << ": " << sizeof(argv[i][j]) << "\n";
    //     }
    // }

    // Analyse der Befehlszeilenoptionen
    if (!parseOptions(argc, argv, binary, octal, hex, decimal))
    {
        return 1; // Beende das Programm mit Fehlercode 1 bei einem Analysefehler
    }

    // Falls keine Basisoptionen angegeben sind, verwenden Sie alle drei
    if (!binary && !octal && !hex)
    {
        binary = octal = hex = true;
    }

    // std::cout << "Dual: " << binary << "\nOctal: " << octal << "\nHex: " << hex << "\n";

    // Anzeige der Ergebnisse
    printResults(binary, octal, hex, decimal);

    return 0; // Beende das Programm ohne Fehler
}
